ASoC: Realtek codecs: wait codec init in hw_params

bardliao opened the pull request:
Try to reduce the resume time

bardliao committed:
ASoC: rt722: wait codec init in hw_params

Move regmap sync to rt722_sdca_update_status() when unattach_request
is set, and only do regmap sync in resume when no reattach needed.
And move waiting codec init to hw_params when the codec really need to
be initialized. The change can shorten the resume time.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: rt712: wait codec init in hw_params

Move regmap sync to rt712_sdca_update_status() when unattach_request
is set, and only do regmap sync in resume when no reattach needed.
And move waiting codec init to hw_params when the codec really need to
be initialized. The change can shorten the resume time.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: rt1320: wait codec init in hw_params

Move regmap sync to rt1320_update_status() when unattach_request
is set, and only do regmap sync in resume when no reattach needed.
And move waiting codec init to hw_params when the codec really need to
be initialized. The change can shorten the resume time.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao commented:
`rt722-sdca sdw:0:3:025d:0722:01: SDW_SCP_BUSCLOCK_SCALE register write failed` is reported. https://sof-ci.01.org/linuxpr/PR5424/build7447/devicetest/index.html. But I can't reproduce the issue. Rerun CI test.

bardliao commented:
SOFCI TEST

bardliao commented:
SOFCI TEST

charleskeepax commented on drivers/soundwire/stream.c:
```diff
@@ -2134,9 +2136,19 @@ int sdw_stream_add_slave(struct sdw_slave *slave,
+		sdw_show_ping_status(slave->bus, true);
+
+		return -ETIMEDOUT;
+	}
```
Definitely have some concerns about moving this to the core, could we do this in the hw_params of the affected drivers instead? Not all drivers use this initialization complete stuff, and this will have impacts across everything.

bardliao replied:
Yes, we could move this to the hw_params of each codec drivers. But given that `complete_all(&slave->initialization_complete);` is done in the core, and the codec should be initialized (or at least be attached) when it is added to the stream, it also makes sense to wait here, right? Would it harm in some corner case?

bardliao replied:
BTW, I will remove `wait_for_completion_timeout(&slave->initialization_complete` for all codec drivers if we go with this commit. Right now I just do it for the 3 Realtek codec drivers because the 3 codecs are used by our CI machines. And I want to let CI confirm that I didn't break anything with the change.

charleskeepax replied:
My problem really is that I am not sure that initialization_complete bit should live in the core at all (well unless we did something more detailed with actually probing after the device has shown up on the bus). Different drivers are likely to want to handle that initialization stuff in different ways so I am uncomfortable with it being in the core. Perhaps we could take an approach rather than adding it to the add_slave callback make it a separate helper function? That way you could still save duplicating the code in many codec drivers, but without forcing every codec to run that code?

bardliao committed:
soundwire: add sdw_slave_wait_for_initialization helper

waiting for slave->initialization_complete is commonly used by SoundWire
codec drivers. Add a helper to reduce the duplicated code.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: rt721: wait codec init in hw_params

Move regmap sync to rt721_sdca_update_status() when unattach_request
is set, and only do regmap sync in resume when no reattach needed.
And move waiting codec init to hw_params when the codec really need to
be initialized. The change can shorten the resume time.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: rt715-sdca: wait codec init in hw_params

Move regmap sync to rt715_sdca_update_status() when unattach_request
is set, and only do regmap sync in resume when no reattach needed.
And move waiting codec init to hw_params when the codec really need to
be initialized. The change can shorten the resume time.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: rt711-sdca: wait codec init in hw_params

Move regmap sync to rt711_sdca_update_status() when unattach_request
is set, and only do regmap sync in resume when no reattach needed.
And move waiting codec init to hw_params when the codec really need to
be initialized. The change can shorten the resume time.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: rt711: wait codec init in hw_params

Move regmap sync to rt711_update_status() when unattach_request
is set, and only do regmap sync in resume when no reattach needed.
And move waiting codec init to hw_params when the codec really need to
be initialized. The change can shorten the resume time.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: rt715: wait codec init in hw_params

Move regmap sync to rt715_update_status() when unattach_request
is set, and only do regmap sync in resume when no reattach needed.
And move waiting codec init to hw_params when the codec really need to
be initialized. The change can shorten the resume time.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: rt700: wait codec init in hw_params

Move regmap sync to rt700_update_status() when unattach_request
is set, and only do regmap sync in resume when no reattach needed.
And move waiting codec init to hw_params when the codec really need to
be initialized. The change can shorten the resume time.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: rt1316: wait codec init in hw_params

Move regmap sync to rt1316_update_status() when unattach_request
is set, and only do regmap sync in resume when no reattach needed.
And move waiting codec init to hw_params when the codec really need to
be initialized. The change can shorten the resume time.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: rt1318: wait codec init in hw_params

Move regmap sync to rt1318_update_status() when unattach_request
is set, and only do regmap sync in resume when no reattach needed.
And move waiting codec init to hw_params when the codec really need to
be initialized. The change can shorten the resume time.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: rt1308: wait codec init in hw_params

Move regmap sync to rt1308_update_status() when unattach_request
is set, and only do regmap sync in resume when no reattach needed.
And move waiting codec init to hw_params when the codec really need to
be initialized. The change can shorten the resume time.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: rt5682: wait codec init in hw_params

Move regmap sync to rt5682_update_status() when unattach_request
is set, and only do regmap sync in resume when no reattach needed.
And move waiting codec init to hw_params when the codec really need to
be initialized. The change can shorten the resume time

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: rt1017: wait codec init in hw_params

Move regmap sync to rt1017_sdca_update_status() when unattach_request
is set, and only do regmap sync in resume when no reattach needed.
And move waiting codec init to hw_params when the codec really need to
be initialized. The change can shorten the resume time.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao commented:
SOFCI TEST

bardliao commented:
SOFCI TEST
